[SPARK-24865] Remove AnalysisBarrier

rxin opened the pull request:
## What changes were proposed in this pull request?
AnalysisBarrier was introduced in SPARK-20392 to improve analysis speed (don't re-analyze nodes that have already been analyzed).

Before AnalysisBarrier, we already had some infrastructure in place, with analysis specific functions (resolveOperators and resolveExpressions). These functions do not recursively traverse down subplans that are already analyzed (with a mutable boolean flag _analyzed). The issue with the old system was that developers started using transformDown, which does a top-down traversal of the plan tree, because there was not top-down resolution function, and as a result analyzer performance became pretty bad.

In order to fix the issue in SPARK-20392, AnalysisBarrier was introduced as a special node and for this special node, transform/transformUp/transformDown don't traverse down. However, the introduction of this special node caused a lot more troubles than it solves. This implicit node breaks assumptions and code in a few places, and it's hard to know when analysis barrier would exist, and when it wouldn't. Just a simple search of AnalysisBarrier in PR discussions demonstrates it is a source of bugs and additional complexity.

Instead, this pull request removes AnalysisBarrier and reverts back to the old approach. We added infrastructure in tests that fail explicitly if transform methods are used in the analyzer.

## How was this patch tested?
Added a test suite AnalysisHelperSuite for testing the resolve* methods and transform* methods.

rxin committed:
[SPARK-24865] Remove AnalysisBarrier

SparkQA commented:
**[Test build #93306 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93306/testReport)** for PR 21822 at commit [`f6f2bcc`](https://github.com/apache/spark/commit/f6f2bccd6d887b50b035c492c88e9e76d1ef4754).

rxin committed:
Minimize change

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1147/
Test PASSed.

SparkQA commented:
**[Test build #93307 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93307/testReport)** for PR 21822 at commit [`8ccafca`](https://github.com/apache/spark/commit/8ccafcab20a70df7a625912fdf4e43be7fb87954).

rxin committed:
Fix bug

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1148/
Test PASSed.

SparkQA commented:
**[Test build #93308 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93308/testReport)** for PR 21822 at commit [`0afa7ea`](https://github.com/apache/spark/commit/0afa7eadb6ae6b97e28eae93d6d2eb83b1f930fd).

SparkQA commented:
**[Test build #93306 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93306/testReport)** for PR 21822 at commit [`f6f2bcc`](https://github.com/apache/spark/commit/f6f2bccd6d887b50b035c492c88e9e76d1ef4754).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93306/
Test FAILed.

SparkQA commented:
**[Test build #93307 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93307/testReport)** for PR 21822 at commit [`8ccafca`](https://github.com/apache/spark/commit/8ccafcab20a70df7a625912fdf4e43be7fb87954).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93307/
Test FAILed.

SparkQA commented:
**[Test build #93308 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93308/testReport)** for PR 21822 at commit [`0afa7ea`](https://github.com/apache/spark/commit/0afa7eadb6ae6b97e28eae93d6d2eb83b1f930fd).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93308/
Test FAILed.

rxin committed:
More bug fixes

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1150/
Test PASSed.

SparkQA commented:
**[Test build #93310 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93310/testReport)** for PR 21822 at commit [`738e99c`](https://github.com/apache/spark/commit/738e99c615f45cfb6e5882a822dafcb8472b78ea).

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -231,10 +231,11 @@ class Analyzer(
       // Lookup WindowSpecDefinitions. This rule works with unresolved children.
       case WithWindowDefinition(windowDefinitions, child) =>
-        child.transform {
+        // TODO(rxin): Check with Herman whether the next line is OK.
```
cc @hvanhovell

hvanhovell replied:
It is good. The earlier `resolveOperators` makes sure we don't overwrite a window spec, with a similarly named one defined higher up the tree. BTW I don't think we have a test that covers this (it is pretty rare).

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -787,14 +782,17 @@ class Analyzer(
-                dedupAttr(a, attributeRewrites)
-              case s: SubqueryExpression =>
-                s.withNewPlan(dedupOuterReferencesInSubquery(s.plan, attributeRewrites))
+          // TODO(rxin): Why do we need transformUp here?
```
cc @hvanhovell @cloud-fan

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -533,7 +537,8 @@ trait CheckAnalysis extends PredicateHelper {
     // Simplify the predicates before validating any unsupported correlation patterns
     // in the plan.
-    BooleanSimplification(sub).foreachUp {
+    // TODO(rxin): Why did this need to call BooleanSimplification???
```
cc @dilipbiswal @cloud-fan @gatorsmile 

Why did we need BooleanSimplification here?

dilipbiswal replied:
@rxin From what i remember Reynold, most of this logic was housed in Analyzer before and we moved it to optimizer. In the old code we used to walk the plan after simplifying the predicates. The comment used to read "Simplify the predicates before pulling them out.". I just retained that semantics.

dilipbiswal replied:
@rxin I tracked down the PR that introduced the change in the Analyzer. Here is the link -
https://github.com/apache/spark/pull/12954

hvanhovell replied:
Yeah, I added boolean simplification here. I didn't quite like it back then, and I still don't like it. I was hoping this was happening in the `Optimizer` now.

dilipbiswal replied:
@hvanhovell Hi Herman, as you said, we do the actual pulling up of the predicates in the optimizer in PullupCorrelatedPredicates in subquery.scala. We are also doing a BooleanSimplication first before traversing the plan there. In here, we are doing the error reporting and i thought it would be better to keep the traversal the same way. Basically previously we did the error reporting and rewriting in Analyzer and now, we do the error reporting in checkAnalysis and rewriting in Optimizer. Just to refresh your memory so you can help to take the right call here :-)

hvanhovell replied:
Well tests fail without it, so we don't really have a choice here. For a second I thought we could also create some utils class, but that would just mean moving the code in BooleanSimplification in there just for esthetics.

dilipbiswal replied:
@hvanhovell Yeah. I agree.

SparkQA commented:
**[Test build #93310 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93310/testReport)** for PR 21822 at commit [`738e99c`](https://github.com/apache/spark/commit/738e99c615f45cfb6e5882a822dafcb8472b78ea).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93310/
Test FAILed.

rxin committed:
more fixes

SparkQA commented:
**[Test build #93320 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93320/testReport)** for PR 21822 at commit [`83ffa51`](https://github.com/apache/spark/commit/83ffa51f4b165152dea214be4d73dd518d742a56).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1156/
Test PASSed.

SparkQA commented:
**[Test build #93320 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93320/testReport)** for PR 21822 at commit [`83ffa51`](https://github.com/apache/spark/commit/83ffa51f4b165152dea214be4d73dd518d742a56).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93320/
Test FAILed.

maropu commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1158/
Test PASSed.

SparkQA commented:
**[Test build #93325 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93325/testReport)** for PR 21822 at commit [`83ffa51`](https://github.com/apache/spark/commit/83ffa51f4b165152dea214be4d73dd518d742a56).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93325/
Test FAILed.

rxin committed:
bypass EliminateSubqueryAliases

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1175/
Test PASSed.

SparkQA commented:
**[Test build #93351 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93351/testReport)** for PR 21822 at commit [`7c76c83`](https://github.com/apache/spark/commit/7c76c83fe89f3e5aa28540fd76bdfc6016c35749).

SparkQA commented:
**[Test build #93351 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93351/testReport)** for PR 21822 at commit [`7c76c83`](https://github.com/apache/spark/commit/7c76c83fe89f3e5aa28540fd76bdfc6016c35749).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93351/
Test FAILed.

rxin committed:
Added BooleanSimplification back

rxin committed:
revert mistake

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1187/
Test PASSed.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -2390,16 +2375,21 @@ class Analyzer(
-    case SubqueryAlias(_, child) => child
+  // This is actually called in the beginning of the optimization phase, and as a result
+  // is using transformUp rather than resolveOperators. This is also often called in the
+  //
```
note: finish comment

SparkQA commented:
**[Test build #93366 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93366/testReport)** for PR 21822 at commit [`38980ad`](https://github.com/apache/spark/commit/38980ad066d26327387673910e0dfd981102cab9).

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:
```diff
@@ -23,8 +23,24 @@ import org.apache.spark.sql.catalyst.analysis._
+
+object LogicalPlan {
+
+  private val resolveOperatorDepth = new ThreadLocal[Int] {
```
todo: explain what this is

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:
```diff
@@ -33,6 +49,116 @@ abstract class LogicalPlan
+   *
+   * @param rule the function use to transform this nodes children
+   */
+  def resolveOperators(rule: PartialFunction[LogicalPlan, LogicalPlan]): LogicalPlan = {
```
todo: add unit tests

rxin commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1190/
Test PASSed.

SparkQA commented:
**[Test build #93370 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93370/testReport)** for PR 21822 at commit [`38980ad`](https://github.com/apache/spark/commit/38980ad066d26327387673910e0dfd981102cab9).

SparkQA commented:
**[Test build #93366 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93366/testReport)** for PR 21822 at commit [`38980ad`](https://github.com/apache/spark/commit/38980ad066d26327387673910e0dfd981102cab9).
 * This patch **fails from timeout after a configured wait of \`300m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93366/
Test FAILed.

rxin commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1195/
Test PASSed.

SparkQA commented:
**[Test build #93375 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93375/testReport)** for PR 21822 at commit [`38980ad`](https://github.com/apache/spark/commit/38980ad066d26327387673910e0dfd981102cab9).

SparkQA commented:
**[Test build #93370 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93370/testReport)** for PR 21822 at commit [`38980ad`](https://github.com/apache/spark/commit/38980ad066d26327387673910e0dfd981102cab9).
 * This patch **fails from timeout after a configured wait of \`300m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93370/
Test FAILed.

SparkQA commented:
**[Test build #93375 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93375/testReport)** for PR 21822 at commit [`38980ad`](https://github.com/apache/spark/commit/38980ad066d26327387673910e0dfd981102cab9).
 * This patch **fails from timeout after a configured wait of \`300m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93375/
Test FAILed.

HyukjinKwon commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1197/
Test PASSed.

SparkQA commented:
**[Test build #93380 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93380/testReport)** for PR 21822 at commit [`38980ad`](https://github.com/apache/spark/commit/38980ad066d26327387673910e0dfd981102cab9).

HyukjinKwon commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1205/
Test PASSed.

SparkQA commented:
**[Test build #93391 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93391/testReport)** for PR 21822 at commit [`38980ad`](https://github.com/apache/spark/commit/38980ad066d26327387673910e0dfd981102cab9).

SparkQA commented:
**[Test build #93391 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93391/testReport)** for PR 21822 at commit [`38980ad`](https://github.com/apache/spark/commit/38980ad066d26327387673910e0dfd981102cab9).
 * This patch **fails from timeout after a configured wait of \`300m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93391/
Test FAILed.

kiszk commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1214/
Test PASSed.

SparkQA commented:
**[Test build #93403 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93403/testReport)** for PR 21822 at commit [`38980ad`](https://github.com/apache/spark/commit/38980ad066d26327387673910e0dfd981102cab9).

SparkQA commented:
**[Test build #93403 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93403/testReport)** for PR 21822 at commit [`38980ad`](https://github.com/apache/spark/commit/38980ad066d26327387673910e0dfd981102cab9).
 * This patch **fails from timeout after a configured wait of \`300m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93403/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #93420 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93420/testReport)** for PR 21822 at commit [`38980ad`](https://github.com/apache/spark/commit/38980ad066d26327387673910e0dfd981102cab9).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1224/
Test PASSed.

SparkQA commented:
**[Test build #93420 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93420/testReport)** for PR 21822 at commit [`38980ad`](https://github.com/apache/spark/commit/38980ad066d26327387673910e0dfd981102cab9).
 * This patch **fails from timeout after a configured wait of \`300m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93420/
Test FAILed.

HyukjinKwon commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1252/
Test PASSed.

SparkQA commented:
**[Test build #93470 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93470/testReport)** for PR 21822 at commit [`38980ad`](https://github.com/apache/spark/commit/38980ad066d26327387673910e0dfd981102cab9).

rxin commented:
Yea the extra check in test cases might've contributed to the longer test
time. Let me think about how to reduce it.


On Mon, Jul 23, 2018 at 11:28 PM Hyukjin Kwon <notifications@github.com>
wrote:

> @rxin <https://github.com/rxin>, I think this PR could possibly cause
> some performance effect given the latest test ran above
> https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93470/
> and from a rough scan of other builds:
>
> ArithmeticExpressionSuite:
> - SPARK-22499: Least and greatest should not generate codes beyond 64KB (11 minutes, 51 seconds)
>
> CastSuite:
> - cast string to timestamp (8 minutes, 42 seconds)
>
> TPCDSQuerySuite:
> - q14a-v2.7 (2 minutes, 29 seconds)
>
> SQLQueryTestSuite:
> - subquery/in-subquery/in-joins.sql (2 minutes, 36 seconds)
>
> ContinuousStressSuite:
> - only one epoch (3 minutes, 21 seconds)
> - automatic epoch advancement (3 minutes, 21 seconds)
>
> vs
> https://amplab.cs.berkeley.edu/jenkins/job/spark-master-test-sbt-hadoop-2.7/4699/consoleFull
>
>
> ArithmeticExpressionSuite:
> - SPARK-22499: Least and greatest should not generate codes beyond 64KB (7 minutes, 49 seconds)
>
> CastSuite:
> - cast string to timestamp (1 minute)
>
> TPCDSQuerySuite:
> - q14a-v2.7 (3 seconds, 442 milliseconds)
>
> SQLQueryTestSuite:
> - subquery/in-subquery/in-joins.sql (2 minutes, 21 seconds)
>
> ContinuousStressSuite:
> - only one epoch (3 minutes, 21 seconds)
> - automatic epoch advancement (3 minutes, 21 seconds)
>
> There could of course other factors like machine's status as well but I
> thought it's good to note while I am taking a look for those.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/spark/pull/21822#issuecomment-407295739>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AATvPMFWaX8huI8hkWBghiP740jtu592ks5uJr6fgaJpZM4VXRDF>
> .
>

SparkQA commented:
**[Test build #93470 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93470/testReport)** for PR 21822 at commit [`38980ad`](https://github.com/apache/spark/commit/38980ad066d26327387673910e0dfd981102cab9).
 * This patch **fails from timeout after a configured wait of \`300m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93470/
Test FAILed.

rxin committed:
Switch to use thread local to make things go faster.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1293/
Test PASSed.

SparkQA commented:
**[Test build #93524 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93524/testReport)** for PR 21822 at commit [`abfd0a8`](https://github.com/apache/spark/commit/abfd0a8cd16c54fa19e7c66bfc2fb1f1c6b85a12).

rxin committed:
Test cases

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -787,6 +782,7 @@ class Analyzer(
           right
         case Some((oldRelation, newRelation)) =>
           val attributeRewrites = AttributeMap(oldRelation.output.zip(newRelation.output))
+          // TODO(rxin): Why do we need transformUp here?
```
@cloud-fan  ?

cloud-fan replied:
we still need to transform resolved plan here to resolve self-join. Image
```
val df = ...
df.as("a").join(df.as("b"), ...)
```
We need to look into the resolved plan to replace the conflicted attributes.

rxin committed:
Remove one TODO

SparkQA commented:
**[Test build #93526 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93526/testReport)** for PR 21822 at commit [`75fb114`](https://github.com/apache/spark/commit/75fb1145fa84725fb931752906ece34ae6028290).

rxin commented:
I changed the way we do the checks in test to use a thread local rather than checking the stacktrace, so they should run faster now. Also added test cases for the various new methods. Also moved the relevant code into AnalysisHelper for better code structure.

This should be ready now if tests pass.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1295/
Test PASSed.

SparkQA commented:
**[Test build #93527 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93527/testReport)** for PR 21822 at commit [`f2f1a97`](https://github.com/apache/spark/commit/f2f1a97e447a41e8b9b6c094376d32b32af00991).

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -751,7 +751,8 @@ object TypeCoercion {
 
-    override protected def coerceTypes(plan: LogicalPlan): LogicalPlan = plan transform { case p =>
+    override protected def coerceTypes(
+      plan: LogicalPlan): LogicalPlan = plan resolveOperatorsDown { case p =>
```
im using a weird wrapping here to minimize the diff.

SparkQA commented:
**[Test build #93524 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93524/testReport)** for PR 21822 at commit [`abfd0a8`](https://github.com/apache/spark/commit/abfd0a8cd16c54fa19e7c66bfc2fb1f1c6b85a12).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait AnalysisHelper extends QueryPlan[LogicalPlan] `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93524/
Test FAILed.

gatorsmile commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1300/
Test PASSed.

SparkQA commented:
**[Test build #93533 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93533/testReport)** for PR 21822 at commit [`f2f1a97`](https://github.com/apache/spark/commit/f2f1a97e447a41e8b9b6c094376d32b32af00991).

SparkQA commented:
**[Test build #93526 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93526/testReport)** for PR 21822 at commit [`75fb114`](https://github.com/apache/spark/commit/75fb1145fa84725fb931752906ece34ae6028290).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93526/
Test PASSed.

SparkQA commented:
**[Test build #93533 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93533/testReport)** for PR 21822 at commit [`f2f1a97`](https://github.com/apache/spark/commit/f2f1a97e447a41e8b9b6c094376d32b32af00991).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93533/
Test PASSed.

gatorsmile committed:
Merge remote-tracking branch 'upstream/master' into SPARK-24865-xiao

gatorsmile committed:
add a dummy fix

rxin committed:
Merge pull request #22 from gatorsmile/SPARK-24865-xiao

resolve the conflicts

SparkQA commented:
**[Test build #93630 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93630/testReport)** for PR 21822 at commit [`fe52801`](https://github.com/apache/spark/commit/fe528010f98ccbafaa486a8d57afced0e6f10393).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93630/
Test FAILed.

gatorsmile commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1388/
Test PASSed.

SparkQA commented:
**[Test build #93648 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93648/testReport)** for PR 21822 at commit [`fe52801`](https://github.com/apache/spark/commit/fe528010f98ccbafaa486a8d57afced0e6f10393).

SparkQA commented:
**[Test build #93648 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93648/testReport)** for PR 21822 at commit [`fe52801`](https://github.com/apache/spark/commit/fe528010f98ccbafaa486a8d57afced0e6f10393).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93648/
Test PASSed.

cloud-fan commented:
LGTM, merging to master!
